packages/ckeditor5-engine/src/model/differ.ts

```diff
@@ -473,18 +473,20 @@ export default class Differ {
 			let j = 0; // Iterator in `snapshotChildren` array -- iterates through old children of element.
 
 			// Process every action.
-			for ( const action of actions ) {
-				if ( action === 'i' ) {
+			for ( const [ action, name ] of actions ) {
+				if ( name === 'i' ) {
 					// Generate diff item for this element and insert it into the diff set.
-					diffSet.push( this._getInsertDiff( element, i, elementChildren[ i ] ) );
+					const maybeChildElementSnapshot = snapshotChildren.find( snapshot => snapshot.node === elementChildren[ i ].node );
```

Comment:
If I am not mistaken, this will work well only if the renamed/refreshed element is still in the same parent.

I wonder if we won't get an incorrect result in following case:

1.  Move element from one parent to another.
2.  Rename that element after moving.

I am afraid, we will get `action: 'rename'` but no `before`. This is quite an edge case, but could happen, and in this part of the editor, we need to consider every edge case (e.g. we don't know what kind of a crazy feature an external integrator will come up with :D, or how operations will get batched in RTC).

If I am right and there's a problem, then there are two ways we can go with it:

1.  Quick and a little dirty one: make sure that `action` is not being set if `before` is not available. This way we won't lead to an incorrect situation. We simply make a decision that if an element was moved at some point, it won't be treated as a renamed element.
2.  Proper: as we discussed, if element got snapshoted at any point, that snapshot should be made available through that element reference.  
      
    Right now, `_elementSnapshots` map is IMO misleading, because it does not keep element => element\_snapshot link, it keeps element => element\_children\_snapshots link. I think we could rename it to `_elementChildrenSnapshots` and then we can have `_elementSnapshots` for element => element\_snapshot links. Then, hopefully we could rewrite line 479 to: `this._elementSnapshots.get( elementChildren[ i ].node )`. WDYT?
    
    Alternatively, we could go through all the snapshots be iterating through everything that's in `_elementSnapshots`. We could even do that preemptively, before starting the process, to evaluate element => element_snapshot pairs without introducing a new structure. That of course adds extra computing but I guess it shouldn't be a problem performance wise. WDYT?

Reply:
I had already tested the second solution before issuing the PR, and because of the larger amount of code, I ended up using the `find` approach. I didn't take into account that edge case you are talking about, and I think it will be safest to simply implement both solutions. I have updated the PR by adding these two approaches.